### PR TITLE
Power on

### PR DESCRIPTION
**Pull request by Izhmash**
Adds the ability to directly power a node on.

### CODE REVIEW

File: daemon.go [hunk 1]
```diff
@@ -95,6 +95,16 @@ func (d *Daemon) DialNodeConsole(label string, token *Token) (io.ReadCloser, err
...
 }
 
+func (d *Daemon) PowerOnNode(label string, token *Token) error {
+	d.Lock()
+	defer d.Unlock()
+	node, err := d.getNodeWithToken(label, token)
+	if err != nil {
+		return err
+	}
+	return node.OBM.PowerOn()
```
Comment (zenhack) [inline daemon.go]: I think it's probably time to factor this pattern out;  we've got the same copypasta across three methods now. Maybe something like:

```go
func (d *Daemon) PowerOnNode(label string, token *Token) error {
        d.usingNodeWithToken(label, token, func(n *Node) error {
                return n.OBM.PowerOn()
        }
}
```

with an appropriate definition of `usingNodeWithToken`.

**Comment by Izhmash** [comment]
@zenhack what do you think of the solution I posted?

File: daemon.go [hunk 2]
```diff
@@ -95,24 +107,34 @@ func (d *Daemon) DialNodeConsole(label string, token *Token) (io.ReadCloser, err
 	return node.OBM.DialConsole()
 }
 
+func (d *Daemon) PowerOnNode(label string, token *Token) error {
+	err := d.usingNodeWithToken(label, token, func(n *Node) error {
+		return n.OBM.PowerOn()
```
Comment (naved001) [inline daemon.go]: took me way longer than I'd like to admit to wrap my head around this syntax; still getting used to it.

File: daemon.go [hunk 3]
```diff
@@ -85,6 +85,18 @@ func (d *Daemon) getNodeWithToken(label string, token *Token) (*Node, error) {
...
+func (d *Daemon) usingNodeWithToken(label string, token *Token,
+	f func(*Node) error) error {
+	d.Lock()
+	defer d.Unlock()
+	node, err := d.getNodeWithToken(label, token)
+	if err != nil {
+		return err
+	}
+	f(node)
+	return nil
```
Comment (zenhack) [inline daemon.go]: Should change this to `return f(node)`; otherwise we're silencing errors from f.

File: daemon.go [hunk 4]
```diff
@@ -95,24 +107,34 @@ func (d *Daemon) DialNodeConsole(label string, token *Token) (io.ReadCloser, err
...
 }
 
+func (d *Daemon) PowerOnNode(label string, token *Token) error {
+	err := d.usingNodeWithToken(label, token, func(n *Node) error {
+		return n.OBM.PowerOn()
+	})
+	if err != nil {
+		return err
+	}
+	return nil
```
Comment (zenhack) [inline daemon.go]: Why not just `return d.usingNodeWithToken(...) {`?

File: daemon.go (hunk 3 above)
Comment (naved001) [reply daemon.go]: good catch, in my head I was thinking that `f(node)` will "raise" an error rather than simply "return" error

**Comment by zenhack** [comment]
Quoting Naved Ansari (2018-05-29 14:47:55)

>    good catch, in my head I was thinking that f(node) will "raise" an
>    error rather than simply "return" error

Makes sense. Go doesn't really have exceptions as such. There's
panic/recover, which are mechanically similar, but the idiom is to use
those in the way we make use of assert in Python -- as a
this-should-never-happen sanity check, not a routine error handling
mechanism; in general, errors are reported via return values.

**Comment by Izhmash** [comment]
That all definitely makes sense, I think I addressed all of the issues.

**Comment by zenhack** [comment]
LGTM, merging.

**Pull request merged**